subsys/net/lib/sockets/sockets.c

```diff
@@ -30,26 +30,43 @@ LOG_MODULE_REGISTER(net_sock, CONFIG_NET_SOCKETS_LOG_LEVEL);
 #define SET_ERRNO(x) \
 	{ int _err = x; if (_err < 0) { errno = -_err; return -1; } }
 
-#define VTABLE_CALL(fn, sock, ...) \
-	do { \
-		const struct socket_op_vtable *vtable; \
-		void *ctx = get_sock_vtable(sock, &vtable); \
-		if (ctx == NULL || vtable->fn == NULL) { \
-			errno = EBADF; \
-			return -1; \
-		} \
-		return vtable->fn(ctx, __VA_ARGS__); \
+#define VTABLE_CALL(fn, sock, ...)			     \
+	do {						     \
+		const struct socket_op_vtable *vtable;	     \
+		struct socket_obj_locking obj_locking;	     \
+		struct sys_mutex *lock;			     \
+		void *obj;				     \
+		int ret;				     \
+							     \
+		obj = get_sock_vtable(sock, &vtable, &lock); \
+		if (obj == NULL || vtable->fn == NULL) {     \
+			errno = EBADF;			     \
+			return -1;			     \
+		}					     \
+							     \
+		obj_locking.obj = obj;			     \
```

Comment:
Again, a very unlikely scenario, but if we get preempted after `get_sock_vtable` and before `sys_mutex_lock` and the socket gets closed, we again might use the mutex we no longer own. Perhaps the scheduler shall be locked before `get_sock_vtable` call to prevent that?

Reply:
I removed the commit that introduced this.